dask/array/random.py

```diff
@@ -17,7 +17,603 @@
+
+    ``Generator`` does not provide a version compatibility guarantee. In
+    particular, as better algorithms evolve the bit stream may change.
+
+    This object is identical to ``np.random.Generator`` except that all
+    functions also take a ``chunks=`` keyword argument.
+
+    Parameters
+    ----------
+    bit_generator : BitGenerator
+        BitGenerator to use as the core generator.
+
+    Notes
+    -----
+    The Python stdlib module `random` contains pseudo-random number generator
+    with a number of methods that are similar to the ones available in
+    ``Generator``. It uses Mersenne Twister, and this bit generator can
+    be accessed using ``MT19937``. ``Generator``, besides being
+    Dask-aware, has the advantage that it provides a much larger number
+    of probability distributions to choose from.
+
+    Examples
+    --------
+    >>> from numpy.random import PCG64
+    >>> from dask.array.random import Generator
+    >>> rng = Generator(PCG64())
+    >>> rng.standard_normal().compute() # doctest: +SKIP
+    array(0.44595957)  # random
+
+    See Also
+    --------
+    default_rng : Recommended constructor for `Generator`.
+    np.random.Generator
+    """
+
+    def __init__(self, bit_generator):
+        self._bit_generator = bit_generator
+        self._bit_generator_state = bit_generator.state
+        self._generator = Generator
+
+    def __str__(self):
+        _str = self.__class__.__name__
+        _str += "(" + self._bit_generator.__class__.__name__ + ")"
+        return _str
+
+    def _wrap(
+        self, funcname, *args, size=None, chunks="auto", extra_chunks=(), **kwargs
+    ):
+        """Wrap numpy random function to produce dask.array random function
+
+        extra_chunks should be a chunks tuple to append to the end of chunks
+        """
+        if size is not None and not isinstance(size, (tuple, list)):
+            size = (size,)
+
+        shapes = list(
+            {
+                ar.shape
+                for ar in chain(args, kwargs.values())
+                if isinstance(ar, (Array, np.ndarray))
+            }
+        )
+        if size is not None:
+            shapes.append(size)
+        # broadcast to the final size(shape)
+        size = broadcast_shapes(*shapes)
+        chunks = normalize_chunks(
+            chunks,
+            size,  # ideally would use dtype here
+            dtype=kwargs.get("dtype", np.float64),
+        )
+        slices = slices_from_chunks(chunks)
+
+        def _broadcast_any(ar, shape, chunks):
+            if isinstance(ar, Array):
+                return broadcast_to(ar, shape).rechunk(chunks)
+            if isinstance(ar, np.ndarray):
+                return np.ascontiguousarray(np.broadcast_to(ar, shape))
+
+        # Broadcast all arguments, get tiny versions as well
+        # Start adding the relevant bits to the graph
+        dsk = {}
+        lookup = {}
+        small_args = []
+        dependencies = []
```

Comment:
Why the small versions? Couldn't we just use the `._meta`?

Reply:
same as above. prefer to keep as it is if there arent any strong objections